src/Inspector/utils/inspect.ts

```diff
@@ -28,13 +34,98 @@ export const getElementCodeInfo = (element: HTMLElement): CodeInfo | undefined =
     }
   }
 
+  return undefined
+}
+
+/**
+ * judge if the inspected element is a component
+ */
+const isComponentFiber = (fiber: Fiber) => {
+  if (fiber.index || fiber.sibling) {
+    /**
+     * a special case like:
+     *
+     * const App = () => {
+     *  return (
+     *    <>
+     *      <p> my fiber.sibling is div and my fiber.index = 0 </p>
+     *      <div> my fiber.sibling is null but my fiber.index = 1 </div>
+     *      both of p's and div's father fiber are App's fiber, but they are not components
+     *    </>
+     *  )
+     * }
+     */
+    return false
+  }
+  /**
+   * a actual dom node's return property of fiber points to its father fiber(component's fiber)
+   *
+   * such as: const App = () => <div line="2" />; --> inspected element fiber
+   * <App line="1"/> --> father fiber
+   * component fiber doesn't has the stateNode(actual dom node)
+   */
+  const fatherFiber = fiber?.return
+  return !fatherFiber.stateNode
+}
+
+/**
+ * get code info from normal nodes like div, span
+ */
+export const getCodeInfoFromNodeFiber = (fiber: Fiber): InjectCodeInfo => fiber.pendingProps
+
+/**
+ * fiber?.return points to component fiber,
+ * both pendingProps and memoizedProps record the runtime props
+ */
+export const getCodeInfoFromComponentFiber = (fiber: Fiber): InjectCodeInfo => fiber?.return.pendingProps
+
+export const getElementCodeInfo = (element: HTMLElement): CodeInfo | undefined => {
+  // data attributes auto create by loader in webpack plugin `inspector-loader`
+  if (!element?.dataset) return undefined
+
+  const inspectedFiber: Fiber | null = getElementFiber(element)
+
+  if (!inspectedFiber) return undefined
+
+  const isComponent = isComponentFiber(inspectedFiber)
+  /**
+   * the components and the normal nodes they can't share common logic to get the code info
+   *
+   * components get code info from father fiber
+   * normal nodes get code info from current fiber
+   */
+  const codePositionInfo: InjectCodeInfo = isComponent
+    ? getCodeInfoFromComponentFiber(inspectedFiber)
+    : getCodeInfoFromNodeFiber(inspectedFiber)
+
+  const codeInfo = getCodeInfoFromInjectCodeInfo(codePositionInfo)
```

Comment:
and I'm looking for a way to solve context and ref forward

Reply:
![image](https://user-images.githubusercontent.com/56421380/109931697-1cb69100-7d04-11eb-90ec-99e920efe6cd.png)
i think we can judge a fiber by fiber tag.But react dosen't expose it to developers

Reply:
@nzreal yes, I did it, code see here:

https://github.com/zthxxx/react-dev-inspector/blob/95970d026cc39e6153c1e9b8d38c4ea76418768d/src/Inspector/utils/inspect.ts#L50-L121

preview deploy see below:

![image](https://user-images.githubusercontent.com/15135943/110020219-4901fa00-7d64-11eb-87e4-fef835549b66.png)

Reply:
Wow, it's a better way to solve this problem. You're so familiar with origin code of react. I think i should take some time to read again.